packages/plugin-e2e/src/models/ExplorePage.ts

```diff
@@ -0,0 +1,49 @@
+import { Expect, Locator } from '@playwright/test';
+import { NavigateOptions, PluginTestCtx, RequestOptions } from '../types';
+import { DataSourcePicker } from './DataSourcePicker';
+import { GrafanaPage } from './GrafanaPage';
+import { TimeRange } from './TimeRange';
+
+export class ExplorePage extends GrafanaPage {
+  datasource: DataSourcePicker;
+  timeRange: any;
+  constructor(ctx: PluginTestCtx, expect: Expect<any>) {
+    super(ctx, expect);
+    this.datasource = new DataSourcePicker(ctx, expect);
+    this.timeRange = new TimeRange(ctx, this.expect);
+  }
+
+  async goto(options?: NavigateOptions) {
+    await super.navigate(this.ctx.selectors.pages.Explore.url, options);
+  }
+
+  async getQueryEditorRow(refId: string): Promise<Locator> {
+    //TODO: add new selector and use it in grafana/ui
+    const locator = this.ctx.page.locator('[aria-label="Query editor row"]').filter({
+      has: this.ctx.page.locator(`[aria-label="Query editor row title ${refId}"]`),
+    });
+    await this.expect(locator).toBeVisible();
+    return locator;
+  }
+
+  async runQuery(options?: RequestOptions) {
+    const components = this.ctx.selectors.components;
+    const runButtonTextContent = await this.getByTestIdOrAriaLabel(components.RefreshPicker.runButtonV2).textContent();
+    // if there is a query running, wait for it to finish before starting a new one
+    if (runButtonTextContent === 'Cancel') {
```

Comment:
is there a way with playwright to check if there's a request to this url ongoing? or use the waitForResponse to return if no request is going?

Reply:
Yeah I agree this is not ideal. I'll look into ways of improving this tomorrow.

Reply:
Networking is an area where Playwright falls a bit short IMO. To achieve something like this, we'd need to setup one event listener that keeps track of new requests, and a second one that keeps track of responses. Then you'd need your own mechanism that correlates these two. 

It seems like the sheets data source is triggering a `onRunQuery` after any change in the query editor. For that reason, there may already be an ongoing request when run query method is executed in the test. This is rather unconventional, so after closer consideration I don't think this should be handled in the lib, but in the data source instead. I'm removing this code.

Reply:
sounds good! the current approach will have to work I guess :)